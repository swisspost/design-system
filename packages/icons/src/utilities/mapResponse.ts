import path from 'path';
import { CenshareResultPage, CenshareResult } from '../models/censhare-result-page.model';
import { IIcon } from '../models/icon.model';

const excludedRanges = [[4000, 7999]];

const isExcluded = (icon: IIcon, filters: number[][]): boolean => {
  const name = Number(icon.file.basename);
  let isExcluded = false;

  filters.forEach(([min, max]) => {
    if (min <= name && name <= max) {
      isExcluded = true;
    }
  });

  return isExcluded;
};

/**
 * Parses zenshare results into a useful format
 * @param response Zenshare result page
 * @returns Array of icons
 */
export const formatResponse = (response: CenshareResultPage): Array<IIcon> => {
  return response.result
    .reduce((acc: IIcon[], item: CenshareResult) => {
      const svgVariant = item.variants?.find(variant => variant.mime === 'image/svg+xml');
      
      if (svgVariant) {
        const fileName = path.basename(svgVariant.name);
        const fileExt = path.extname(svgVariant.name);
        const fileBasename = path.basename(fileName, fileExt).replace(/\s/g, ''); // Some of the icons seem to have a whitespace in the name but not in the filepath itself

        const keywords = [
          ...new Set(
            (item.contentInfo?.freeKeywords ?? '')
              .replace(/(\n|\r\n)/g, '')
              .split(/, ?/)
              .filter(keyword => keyword !== 'Piktogramme "Die Post" ab 2017')
          )
        ];
  
        acc.push({
          uuid: item.uuid,
          id: item.id,
          type: item.type,
          typeFilter: item.typeFilter,
          meta: {
            downloadLink: svgVariant.downloadLink,
            businessfield: item.postInfo?.businessfield,
            keywords,
            year: item.postInfo?.year,
          },
          file: {
            mime: svgVariant.mime,
            name: fileName,
            basename: fileBasename,
            ext: fileExt,
            size: svgVariant.size,
          },
          createdAt: typeof item.createdAt === 'string' ? new Date(item.createdAt) : item.createdAt,
          modifiedAt: typeof item.modifiedAt === 'string' ? new Date(item.modifiedAt) : item.modifiedAt,
        });
      }
      return acc;
    }, [])
    .filter(icon => !isExcluded(icon, excludedRanges));
};
