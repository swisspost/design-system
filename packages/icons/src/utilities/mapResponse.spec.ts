import testResult from '../../tests/fixtures/test-result.json';
import { CenshareResultPage } from '../models/censhare-result-page.model';
import { formatResponse } from './mapResponse';

describe('mapResponse', () => {
  it('Should map the result object to a meaningful data structure', () => {
    const mapped = formatResponse(testResult as CenshareResultPage);
    expect(mapped).toMatchSnapshot();
  });
});
