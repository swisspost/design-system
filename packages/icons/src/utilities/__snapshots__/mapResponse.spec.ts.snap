// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapResponse Should map the result object to a meaningful data structure 1`] = `
[
  {
    "createdAt": 2017-11-10T09:43:19.000Z,
    "file": {
      "basename": "1000",
      "ext": ".svg",
      "mime": "image/svg+xml",
      "name": "1000.svg",
      "size": {
        "dpi": 72,
        "height": 0,
        "width": 0,
      },
    },
    "id": 352084,
    "meta": {
      "businessfield": "kommunikation",
      "downloadLink": "https://cdn.post.ch/hcms/v2.0/entity/asset/352084/storage/MzU0NzA1LzAvbWFzdGVy",
      "keywords": [
        "Briefeinwurf",
        "Brief",
        "Kuvert",
        "Briefumschlag",
        "Briefkasten",
        "Briefaufgabe",
        "Boîte aux lettres",
        "Lettre",
        "enveloppe",
        "boîte aux lettres",
        "dépôt de lettres",
        "Letter box",
        "letter",
        "letter cover",
        "envelope",
        "letter box",
        "posting letters",
        "Briefeinwurf",
        "Brief",
        "Kuvert",
        "Briefumschlag",
        "Briefkasten",
        "Briefaufgabe",
        "Boîte aux lettres",
        "Lettre",
        "enveloppe",
        "boîte aux lettres",
        "dépôt de lettres",
        "Letter box",
        "letter",
        "letter cover",
        "envelope",
        "letter box",
        "posting letters",
      ],
      "year": "2017",
    },
    "modifiedAt": 2021-10-18T11:24:43.000Z,
    "type": "picture.pictogram.",
    "typeFilter": "pictograms",
    "uuid": "95965510-c5fb-11e7-a943-005056a94f4a",
  },
  {
    "createdAt": 2017-11-10T09:50:05.000Z,
    "file": {
      "basename": "1001",
      "ext": ".svg",
      "mime": "image/svg+xml",
      "name": "1001.svg",
      "size": {
        "dpi": 72,
        "height": 0,
        "width": 0,
      },
    },
    "id": 352086,
    "meta": {
      "businessfield": "kommunikation",
      "downloadLink": "https://cdn.post.ch/hcms/v2.0/entity/asset/352086/storage/MzU0NzA2LzAvbWFzdGVy",
      "keywords": [
        "Brief",
        "Briefaufgabe",
        "Kuvert",
        "Briefumschlag",
        "Lettre",
        "dépôt de lettres",
        "enveloppe",
        "",
        "Letter",
        "posting letters",
        "envelope",
        "Brief",
        "Briefaufgabe",
        "Kuvert",
        "Briefumschlag",
        "Lettre",
        "dépôt de lettres",
        "enveloppe",
        "",
        "Letter",
        "posting letters",
        "envelope",
      ],
      "year": undefined,
    },
    "modifiedAt": 2021-10-18T11:24:40.000Z,
    "type": "picture.pictogram.",
    "typeFilter": "pictograms",
    "uuid": "8676c2d0-c5fc-11e7-a943-005056a94f4a",
  },
  {
    "createdAt": 2017-11-10T09:50:05.000Z,
    "file": {
      "basename": "1002",
      "ext": ".svg",
      "mime": "image/svg+xml",
      "name": "1002.svg",
      "size": {
        "dpi": 72,
        "height": 32,
        "width": 32,
      },
    },
    "id": 352087,
    "meta": {
      "businessfield": "kommunikation",
      "downloadLink": "https://cdn.post.ch/hcms/v2.0/entity/asset/352087/storage/MzU0NzA3LzAvbWFzdGVy",
      "keywords": [
        "PostCard Creator",
        "Postkarte",
        "Karte",
        "PostCard Creator",
        "Carte postale",
        "carte",
        "PostCard Creator",
        "postcard",
        "card",
        "PostCard Creator",
        "Postkarte",
        "Karte",
        "PostCard Creator",
        "Carte postale",
        "carte",
        "PostCard Creator",
        "postcard",
        "card",
      ],
      "year": undefined,
    },
    "modifiedAt": 2021-10-18T11:24:40.000Z,
    "type": "picture.pictogram.",
    "typeFilter": "pictograms",
    "uuid": "867c1a00-c5fc-11e7-a943-005056a94f4a",
  },
  {
    "createdAt": 2017-11-10T09:50:05.000Z,
    "file": {
      "basename": "1003",
      "ext": ".svg",
      "mime": "image/svg+xml",
      "name": "1003.svg",
      "size": {
        "dpi": 72,
        "height": 32,
        "width": 32,
      },
    },
    "id": 352088,
    "meta": {
      "businessfield": "kommunikation",
      "downloadLink": "https://cdn.post.ch/hcms/v2.0/entity/asset/352088/storage/MzU0NzA4LzAvbWFzdGVy",
      "keywords": [
        "Gratis Poskarte",
        "Geschenkskarte",
        "Postkarte",
        "Karte",
        "Geschenk",
        "Brief",
        "Carte postale gratuite",
        "carte cadeau",
        "Carte postale",
        "carte",
        "cadeau",
        "lettre",
        "Free postcard",
        "gift card",
        "postcard",
        "card",
        "gift",
        "letter",
        "Gratis Poskarte",
        "Geschenkskarte",
        "Postkarte",
        "Karte",
        "Geschenk",
        "Brief",
        "Carte postale gratuite",
        "carte cadeau",
        "Carte postale",
        "carte",
        "cadeau",
        "lettre",
        "Free postcard",
        "gift card",
        "postcard",
        "card",
        "gift",
        "letter",
      ],
      "year": undefined,
    },
    "modifiedAt": 2021-10-18T11:24:41.000Z,
    "type": "picture.pictogram.",
    "typeFilter": "pictograms",
    "uuid": "86819840-c5fc-11e7-a943-005056a94f4a",
  },
  {
    "createdAt": 2017-11-10T09:50:05.000Z,
    "file": {
      "basename": "1004",
      "ext": ".svg",
      "mime": "image/svg+xml",
      "name": "1004.svg",
      "size": {
        "dpi": 72,
        "height": 32,
        "width": 32,
      },
    },
    "id": 352089,
    "meta": {
      "businessfield": "kommunikation",
      "downloadLink": "https://cdn.post.ch/hcms/v2.0/entity/asset/352089/storage/MzU0NzA5LzAvbWFzdGVy",
      "keywords": [
        "Brief/Paket",
        "Brief",
        "Paket",
        "Kuvert",
        "Paketaufgabe",
        "Briefaufgabe",
        "Briefumschlag",
        "versenden",
        "senden",
        "empfangen",
        "Lettre/colis",
        "Lettre colis",
        "enveloppe",
        "dépôt de colis",
        "dépôt de lettres",
        "enveloppe",
        "envoyer",
        "",
        "Letter/parcel",
        "letter",
        "parcel",
        "letter cover",
        "posting parcels",
        "posting letters",
        "envelope",
        "dispatch",
        "send",
        "receive",
        "Brief/Paket",
        "Brief",
        "Paket",
        "Kuvert",
        "Paketaufgabe",
        "Briefaufgabe",
        "Briefumschlag",
        "versenden",
        "senden",
        "empfangen",
        "Lettre/colis",
        "Lettre colis",
        "enveloppe",
        "dépôt de colis",
        "dépôt de lettres",
        "enveloppe",
        "envoyer",
        "",
        "Letter/parcel",
        "letter",
        "parcel",
        "letter cover",
        "posting parcels",
        "posting letters",
        "envelope",
        "dispatch",
        "send",
        "receive",
      ],
      "year": undefined,
    },
    "modifiedAt": 2021-10-18T11:24:41.000Z,
    "type": "picture.pictogram.",
    "typeFilter": "pictograms",
    "uuid": "86862c20-c5fc-11e7-a943-005056a94f4a",
  },
  {
    "createdAt": 2017-11-10T09:50:05.000Z,
    "file": {
      "basename": "1005",
      "ext": ".svg",
      "mime": "image/svg+xml",
      "name": "1005.svg",
      "size": {
        "dpi": 72,
        "height": 32,
        "width": 32,
      },
    },
    "id": 352090,
    "meta": {
      "businessfield": "kommunikation",
      "downloadLink": "https://cdn.post.ch/hcms/v2.0/entity/asset/352090/storage/MzU0NzEwLzAvbWFzdGVy",
      "keywords": [
        "Paket",
        "Paketaufgabe",
        "Geschenk",
        "Schachtel",
        "senden",
        "empfangen",
        "Colis",
        "dépôt de colis",
        "Cadeau",
        "boîte",
        "envoyer",
        "recevoir",
        "Parcel",
        "posting parcels",
        "gift",
        "box",
        "send",
        "receive",
        "Paket",
        "Paketaufgabe",
        "Geschenk",
        "Schachtel",
        "senden",
        "empfangen",
        "Colis",
        "dépôt de colis",
        "Cadeau",
        "boîte",
        "envoyer",
        "recevoir",
        "Parcel",
        "posting parcels",
        "gift",
        "box",
        "send",
        "receive",
      ],
      "year": undefined,
    },
    "modifiedAt": 2021-10-18T11:24:42.000Z,
    "type": "picture.pictogram.",
    "typeFilter": "pictograms",
    "uuid": "868b3530-c5fc-11e7-a943-005056a94f4a",
  },
  {
    "createdAt": 2017-11-10T09:50:05.000Z,
    "file": {
      "basename": "1006",
      "ext": ".svg",
      "mime": "image/svg+xml",
      "name": "1006.svg",
      "size": {
        "dpi": 72,
        "height": 32,
        "width": 32,
      },
    },
    "id": 352091,
    "meta": {
      "businessfield": "kommunikation",
      "downloadLink": "https://cdn.post.ch/hcms/v2.0/entity/asset/352091/storage/MzU0NzExLzAvbWFzdGVy",
      "keywords": [
        "My Post 24",
        "Paket",
        "24h",
        "Paketaufgabe",
        "senden",
        "empfangen",
        "My Post 24",
        "Colis",
        "24h",
        "dépôt de colis",
        "envoyer",
        "recevoir",
        "My Post 24",
        "parcel",
        "24h",
        "sending parcels",
        "send",
        "receive",
        "My Post 24",
        "Paket",
        "24h",
        "Paketaufgabe",
        "senden",
        "empfangen",
        "My Post 24",
        "Colis",
        "24h",
        "dépôt de colis",
        "envoyer",
        "recevoir",
        "My Post 24",
        "parcel",
        "24h",
        "sending parcels",
        "send",
        "receive",
      ],
      "year": "2017",
    },
    "modifiedAt": 2021-10-18T11:16:36.000Z,
    "type": "picture.pictogram.",
    "typeFilter": "pictograms",
    "uuid": "868fc910-c5fc-11e7-a943-005056a94f4a",
  },
  {
    "createdAt": 2017-11-10T09:50:05.000Z,
    "file": {
      "basename": "1007",
      "ext": ".svg",
      "mime": "image/svg+xml",
      "name": "1007.svg",
      "size": {
        "dpi": 72,
        "height": 32,
        "width": 32,
      },
    },
    "id": 352092,
    "meta": {
      "businessfield": "kommunikation",
      "downloadLink": "https://cdn.post.ch/hcms/v2.0/entity/asset/352092/storage/MzU0NzEyLzAvbWFzdGVy",
      "keywords": [
        "Zustellservice",
        "Zusteller",
        "Paket",
        "Brief",
        "Mensch",
        "Empfänger",
        "Sender",
        "Post empfangen",
        "Paket und Brief",
        "Service de distribution",
        "Personnel de distribution",
        "colis",
        "lettre",
        "homme",
        "destinataire",
        "expéditeur",
        "envoyer du courrier",
        "colis et lettre",
        "Delivery service",
        "delivery staff",
        "parcel",
        "letter",
        "human",
        "recipient",
        "sender",
        "receiving mail",
        "parcel and letter",
        "Zustellservice",
        "Zusteller",
        "Paket",
        "Brief",
        "Mensch",
        "Empfänger",
        "Sender",
        "Post empfangen",
        "Paket und Brief",
        "Service de distribution",
        "Personnel de distribution",
        "colis",
        "lettre",
        "homme",
        "destinataire",
        "expéditeur",
        "envoyer du courrier",
        "colis et lettre",
        "Delivery service",
        "delivery staff",
        "parcel",
        "letter",
        "human",
        "recipient",
        "sender",
        "receiving mail",
        "parcel and letter",
      ],
      "year": undefined,
    },
    "modifiedAt": 2021-10-18T11:24:42.000Z,
    "type": "picture.pictogram.",
    "typeFilter": "pictograms",
    "uuid": "869435e0-c5fc-11e7-a943-005056a94f4a",
  },
  {
    "createdAt": 2017-11-10T09:50:05.000Z,
    "file": {
      "basename": "1008",
      "ext": ".svg",
      "mime": "image/svg+xml",
      "name": "1008.svg",
      "size": {
        "dpi": 72,
        "height": 32,
        "width": 32,
      },
    },
    "id": 352093,
    "meta": {
      "businessfield": "kommunikation",
      "downloadLink": "https://cdn.post.ch/hcms/v2.0/entity/asset/352093/storage/MzU0NzEzLzAvbWFzdGVy",
      "keywords": [
        "Briefzentrum",
        "Zustellung",
        "Zuhause",
        "Brief",
        "Kuvert",
        "Briefe empfangen",
        "Briefumschlag",
        "Centre courrier",
        "Distribution",
        "domicile",
        "lettre",
        "enveloppe",
        "recevoir des lettres",
        "enveloppe",
        "Letter center",
        "delivery",
        "home",
        "letter",
        "envelope",
        "receiving letters",
        "letter cover",
      ],
      "year": undefined,
    },
    "modifiedAt": 2021-10-18T11:32:39.000Z,
    "type": "picture.pictogram.",
    "typeFilter": "pictograms",
    "uuid": "869917e0-c5fc-11e7-a943-005056a94f4a",
  },
  {
    "createdAt": 2017-11-10T09:50:05.000Z,
    "file": {
      "basename": "1009",
      "ext": ".svg",
      "mime": "image/svg+xml",
      "name": "1009.svg",
      "size": {
        "dpi": 72,
        "height": 32,
        "width": 32,
      },
    },
    "id": 352094,
    "meta": {
      "businessfield": "kommunikation",
      "downloadLink": "https://cdn.post.ch/hcms/v2.0/entity/asset/352094/storage/MzU0NzE0LzAvbWFzdGVy",
      "keywords": [
        "Paketzentrum",
        "Zustellung",
        "Zuhause",
        "Paket",
        "empfangen",
        "versenden",
        "senden",
        "Centre colis",
        "Distribution",
        "domicile",
        "colis",
        "recevoir",
        "envoyer",
        "expédier",
        "delivery",
        "at home",
        "parcel",
        "receive",
        "dispatch",
        "send",
      ],
      "year": undefined,
    },
    "modifiedAt": 2021-10-18T11:32:38.000Z,
    "type": "picture.pictogram.",
    "typeFilter": "pictograms",
    "uuid": "86a15540-c5fc-11e7-a943-005056a94f4a",
  },
]
`;
